File: Lintcode 148 - Sort Colors.java
Lintcode 148 - Sort Colors.java

class Solution {
    /**
     * @param nums: A list of integer which is 0, 1 or 2 
     * @return: nothing
     */
    public void sortColors(int[] nums) {
        // write your code here
        if(nums == null || nums.length == 0) {
            return;
        }
        
        int left = 0;
        int right = nums.length-1;
        int i = 0;
        while(i <= right) {
            if(nums[i] == 1) {
                i++;
            } else if (nums[i] == 0) {
                swap(nums, i, left);
                left++;
                i++; //可以++，i之前的数都处理过了
            } else { //nums[i] == 2
                swap(nums, i, right);
                right--;
                //i++; 这里不能++，有可能换过来的是2，2还需重走 while 循环
            }
        }
    }
    
    private void swap(int[] nums, int a, int b) {
        int temp = nums[a];
        nums[a] = nums[b];
        nums[b] = temp;
    }
}

File: Lintcode 148 - Sort Colors.cs
using System;
// using System.Collections;
// using System.Collections.Generic;

namespace CSharp
{
    class Program{
        static void Main()
        {
            int[] nums = new int[]{2,1,0,2,1,1,0,1};
            int[] result = sortArray(nums);
            foreach(int i in result)
            {
                Console.WriteLine(i);
            }
        }

        static int[] sortArray(int[] nums)
        {
            int curr = 0;
            int left = 0;
            int right = nums.Length - 1;
            while(curr <= right)
            {
                if(nums[curr] == 0)
                {
                    swap(nums, curr, left);
                    curr++;
                    left++;
                }
                else if(nums[curr] == 1)
                {
                    curr++;
                }
                else
                {
                    swap(nums, curr, right);
                    right--;
                }
            }

            return nums;
        }

        static void swap(int[] nums, int i, int j)
        {
            int temp = nums[i];
            nums[i] = nums[j];
            nums[j] = temp;
        }
    }
}
